File: coral-model/src/main/kotlin/io/coral/contacts/model/repository/impl/BasicRepositoryImpl.java
package io.coral.contacts.model.repository.impl;

import io.coral.contacts.model.domain.AbstractEntity;
import io.coral.contacts.model.repository.BasicRepository;

import javax.annotation.PostConstruct;
import javax.persistence.EntityManager;
import javax.persistence.EntityManagerFactory;
import javax.persistence.Persistence;
import java.util.List;

public class BasicRepositoryImpl implements BasicRepository {
    private EntityManagerFactory emf = null;

    public EntityManager getEntityManager() {
        init();
        return emf.createEntityManager();
    }

    @PostConstruct
    public void init(){
        System.out.println(".............");
        if (emf == null) {
            emf = Persistence.createEntityManagerFactory("customersPU");
        }
    }

    public AbstractEntity add(AbstractEntity entity){
        EntityManager em=null;
        try {
            em = getEntityManager();
            em.getTransaction().begin();
            em.persist(entity);
            em.flush();
            em.getTransaction().commit();
           return entity;
        } catch ( Exception ex) {
            throw ex;
        } finally {
            em.close();
        }
    }

    public AbstractEntity update(AbstractEntity entity){
        EntityManager em=null;
        try {
            em = getEntityManager();
            em.getTransaction().begin();
            em.merge(entity);
            em.getTransaction().commit();
            return entity;
        } catch ( Exception ex) {
            throw ex;
        } finally {
            em.close();
        }
    }

    public AbstractEntity delete(AbstractEntity entity){
        EntityManager em=null;
        try {
            em = getEntityManager();
            em.getTransaction().begin();
           if (em.contains(entity)){
               em.remove(entity);
           }
            em.getTransaction().commit();
            return entity;
        } catch ( Exception ex) {
            throw ex;
        } finally {
            em.close();
        }
    }

    public List<AbstractEntity> findAll(String entity,int offset,int maxSize){
        EntityManager em=null;
        try {
            em = getEntityManager();
            List<AbstractEntity> list=  em.createNamedQuery(entity+".findAll").setFirstResult(offset).setMaxResults(maxSize).getResultList();
            return list;
        } catch ( Exception ex) {
            throw ex;
        } finally {
            em.close();
        }
    }





}

File: coral-model/src/main/kotlin/io/coral/contacts/model/repository/impl/BasicRepositoryImpl.kt
package io.coral.contacts.model.repository.impl

import io.coral.contacts.model.exception.ObjectNotFoundException
import io.coral.contacts.model.repository.BasicRepository
import io.coral.contacts.model.repository.EMF
import io.tech4health.ts.model.domain.AbstractEntity
import javax.persistence.EntityManager

open class BasicRepositoryImpl : BasicRepository {


    val entityManager: EntityManager
        get() = EMF.getEMF()!!.createEntityManager()




    override fun doAdd(entity: AbstractEntity): AbstractEntity {
        var em: EntityManager? = null
        try {
            em = entityManager
            em.transaction.begin()
            em.persist(entity)
            em.flush()
            em.transaction.commit()
            return entity
        } catch (ex: Exception) {
            throw ex
        } finally {
            em!!.close()
        }
    }

    override fun doUpdate(entity: AbstractEntity): AbstractEntity {
        var em: EntityManager? = null
        try {
            em = entityManager
            em.transaction.begin()
            em.merge(entity)
            em.transaction.commit()
            return entity
        } catch (ex: Exception) {
            throw ex
        } finally {
            em!!.close()
        }
    }

    override fun doDelete(entity: AbstractEntity,id:Int): Boolean {
        var em: EntityManager? = null
        try {
            em = entityManager
            em.transaction.begin()
            val toBeRemoved= em.find(entity::class.java,id) ?:throw ObjectNotFoundException()
            em.remove(toBeRemoved)
            em.transaction.commit()
            return true
        } catch (ex: Exception) {
            throw ex
        } finally {
            em!!.close()
        }
    }

    override fun findAll(entity: String, offset: Int, maxSize: Int): List<AbstractEntity> {
        var em: EntityManager? = null
        try {
            em = entityManager
            return em.createNamedQuery(entity +".findAll").setFirstResult(offset).setMaxResults(maxSize).resultList as List<AbstractEntity>
        } catch (ex: Exception) {
            throw ex
        } finally {
            em!!.close()
        }
    }


}
